Pod os field kubelet

ravisantoshgudimetla opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Once #104613 and #104693
merge, we'll have OS field in pod spec. Kubelet will also start reconciling labels.
From then on, Kubelet should start rejecting pods
where pod.Spec.OS and node's OS label won't match
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Kubelet should reject pods whose OS doesn't match the node's OS label.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/pull/2803
```

fedebongio commented:
/remove-sig api-machinery

ehashman commented:
What is the KEP for this?

/label api-review

ravisantoshgudimetla commented:
/remove label api-review

@ehashman - I updated the description with KEP associated. The api review can be done on https://github.com/kubernetes/kubernetes/pull/104693. I picked the commits from that PR to make the testing easier.

derekwaynecarr commented on pkg/kubelet/lifecycle/predicate.go:
```diff
@@ -153,11 +154,46 @@ func (w *predicateAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult
+	if rejectPodAdmissionBasedOnOSField(admitPod, node) {
+		return PodAdmitResult{
+			Admit:   false,
+			Reason:  "PodOSFieldNodeLabelMisMatch",
```
nit: Mismatch and not MisMatch

derekwaynecarr reviewed (changes requested):
one minor nit , this looks generally good after the prereq pr merges.

ravisantoshgudimetla commented:
/hold 

till https://github.com/kubernetes/kubernetes/pull/104613 merges

ravisantoshgudimetla commented:
/retest

derekwaynecarr commented:
/lgtm
/approve

remove the hold after the requisite pr merges (just tagged)

liggitt commented on pkg/kubelet/lifecycle/predicate.go:
```diff
@@ -153,11 +154,46 @@ func (w *predicateAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult
+	}
+	// Node OS label can be used as source of truth as the OS and arch reconciliation would have happened by now
+	if nodeOS, ok := labels[v1.LabelOSStable]; ok {
+		if string(pod.Spec.OS.Name) != nodeOS {
```
don't compare the pod os to the node labels, compare it to runtime.GOOS

liggitt replied:
the pod nodeSelector should be compared to the node labels with the OS/Arch label ensured to match the real runtime.GOOS/GOARCH, but that's distinct from checking the pod.spec.os against runtime.GOOS

ravisantoshgudimetla committed:
[kubelet] Reject pods with OS field mismatch

Once kubernetes#104613 and kubernetes#104693
merge, we'll have OS field in pod spec. Kubelet should start rejecting pods
where pod.Spec.OS and node's OS(using runtime.GOOS) won't match

ravisantoshgudimetla commented:
@liggitt suggested to not rely on featuregate in kubelet to reject pods. He also suggested to add another featuregate for Pod security standards so that they can get merged in this release instead of [waiting till ](https://github.com/kubernetes/kubernetes/pull/105919#issuecomment-952291110)1.26 or 1.27(when `IdentifyPodOS` feature graduates to GA).

ravisantoshgudimetla commented:
/retest

liggitt commented on pkg/kubelet/lifecycle/predicate.go:
```diff
@@ -161,6 +161,14 @@ func (w *predicateAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult
+	if rejectPodAdmissionBasedOnOSField(admitPod) {
+		return PodAdmitResult{
+			Admit:   false,
+			Reason:  "PodOSFieldNodeLabelDoesNotMatch",
```
drop NodeLabel from the reason… we're not comparing to the label

liggitt replied:
pick a reason that will continue to make sense if LCOW is ever supported… maybe `PodOSNotSupported` or something?

ravisantoshgudimetla replied:
The only reason for keeping it is - the user may not know about the OS of the node. Sometimes the only object he/she may have access to is node object.

liggitt commented:
/lgtm

ravisantoshgudimetla commented:
/retest

ravisantoshgudimetla commented:
/hold cancel

/retest
